Lib/test/test_fractions.py

```diff
@@ -346,6 +346,34 @@ def testConversions(self):
 
         self.assertTypedEquals(0.1+0j, complex(F(1,10)))
 
+    def testBoolGuarateesBoolReturn(self):
+        # Ensure that __bool__ is used on numerator which guarantees a bool
+        # return.  See also bpo-39274.
+        class custom_value(int):
+            def __init__(self, value):
+                self.value = value
+
+            def __bool__(self):
+                self.bool_used = True
+                return bool(self.value)
```

Comment:
The bug is that "numerator != 0" returns an object which is not a subtype of bool. Does your test reproduce the bug if you revert the fractions.py change?

Reply:
True, I am testing that `r._numerator.__bool__` is called, which is not quite the same thing of course. Other solutions to achieve the same things (e.g. `bool(a._numerator != 0)` are possible and would fail the test as well).
I can add `__ne__` that does the bad thing, but it would never get called. This forces to adapt the test in case the code is adapted, but without a full fledged numpy integer around, it seems hard to write a generic test that only checks for a `bool` result.

Reply:
Well, tried to make it a bit more waterproof (by resetting `bool_used`), if I can import NumPy – skipping the test when its not available – that may be nicer of course (but I assume that is a no-go). Otherwise it feels like I would have to implement a full fleged `rational` with added strange logic...

Its over engineered, but right now I do not have a better idea TBH (unless we just skip the test...)

Reply:
You cannot use 3rd party modules like numpy in the Python test suite, sadly. (At least, we are trying hard to avoid that.)

So you should try to mimick numpy's bool behavior. Please just implement a __ne__() method which raises an AssertionError. It would make me more comfortable. Or better: implement __eq__ and __lt__, and @ @functools.total_ordering:
https://docs.python.org/dev/library/functools.html

Again, these methods would raise an exception.

Reply:
Next try... Had tried to avoid registering with `numbers.Rational` but there is probably no reason, so now not inheriting from `int` anymore, which is much cleaner (and added your error, instead of ensuring `__bool__` is called).

Overall, seems much better, since now it really is a crippled object that cannot do anything but what it is supposed to do...